include/envoy/server/filter_config.h

```diff
@@ -201,6 +201,17 @@ class FactoryContext : public virtual CommonFactoryContext {
   virtual ProtobufMessage::ValidationVisitor& messageValidationVisitor() PURE;
 };
 
+/**
+ * An implementation of FactoryContext. The life time is no shorter than the created filter chains.
+ * The life time is no longer than the owning listener. It should be used to create
+ * NetworkFilterChain.
+ */
+class FilterChainFactoryContext : public virtual FactoryContext {};
+
+/**
+ * An implementation of FactoryContext. The life time should cover the lifetime of the filter chains
+ * and connections. It can be used to create ListenerFilterChain.
+ */
```

Comment:
Right now only ListenerImpl is the only class that implements ListenerFactoryContext. Given that FilterChainFactoryContext's lifetime is constrained to no longer than the owning listener, do you plan on a subsequent PR to make one that lives _as long_ as the listener? This way we can avoid having to make the listener itself implement all the FactoryContext functions?

I am coming from thinking about this, for context: https://github.com/envoyproxy/envoy/pull/9516#discussion_r363519159

Reply:
> do you plan on a subsequent PR to make one that lives as long as the listener?

@junr03 My following PRs will focus on generating *short* live `FilterChainFactoryContext` by FilterChainManager. The goal is to to add/remove filter chain(as well as filter chain context) during a listener life time.

I didn't understand ApiListener yet. I feel like you want `FilterChainFactoryContext` since you don't really need listener's information. I think we can have another `FilterChainFactoryContext` impl which doesn't depend on `ListenerImpl`. I think we can moving to this direction in the future.

Reply:
Yeah my plan is to create a FilterChainFactoryContext as a member variable in the ApiListenerImpl that way the ApiListenerImpl does not need to implement all the FactoryContext methods.

Alright, I think I can work with this. I can merge once the tests pass. Thanks!